### PR TITLE
3448 world immovables in encyclopedia

### LINKED ISSUE #3448: Improve tribes help for immovables
We should parse the immovable programs to get more information about them into the encyclopedia.

* Worker plants/harvests
* Building places / collects
* Immovable becomes
* Immovable is planted/harvested by

The information might also be useful for the AI for detecting supporting buildings.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2019-06-06T05:47:15Z
 * owner: gunchleoc
 * assignee: gunchleoc
 * the launchpad url was https://bugs.launchpad.net/bugs/1831842

### PR DESCRIPTION
**Pull request by hessenfarmer**
fixes #3448

### CODE REVIEW

**Comment by hessenfarmer** [comment]
this is only a first draft with several drawbacks just to show how it could look like
drawbacks are:
- dead slow due to the iterations 
- hardcoded tribenames not working for addons

**Comment by hessenfarmer** [comment]
ok fixed the performance issue.
@Nordfrees is there any way to get the complete list of tribes including activated addons in lua. If I need to write a hook for it where to get this information from in c++

**Comment by hessenfarmer** [comment]
update:  solved the tribes issue
got new issues:
1. if not all tribes are loaded, all premature steps of a tribes immovable from a tribe not loaded where the most mature step is used by a loaded tribe are displayed ( e.g. all young amazon trees except the old one if amazons are not loaded) 
2. images are not displayed for most of the world immovables.

possible solutions:
1a. if a tribe is not present do not show any of its immovables at all (neither usable ones in other tribes immovables nor world immovables)
1b.  change handling of co-used immovables to only display them only in the original tribes help.
1c. change definition of world to be everything not defined by own tribe. (results in more world entries as every other tribes immovables will be shown there

2a. create menu icons for each immovable and define them in the tribes descriptions.
2b. use idle.png or other present pictures. 

i tend to favor 1c and 2a

@Noordfrees @frankystone @klaus-halfmann any ideas or opinions?

**Comment by klaus-halfmann** [comment]
> @Noordfrees @frankystone @klaus-halfmann any ideas or opinions?

Hmm, I dont know what the original Problem was, so no idea what I could check.

**Comment by hessenfarmer** [comment]
> 
> 
> > @Noordfrees @frankystone @klaus-halfmann any ideas or opinions?
> 
> Hmm, I dont know what the original Problem was, so no idea what I could check.     
     
this is meant as an enhancement to our in game encyclopedia. It introduces a new tab for world immovables. containing all loaded world immovables and available help for them. 
to see how it looks like compile this and open the in game help and then switch tom the last tab
to see the effects I mentioned do this in agame with only one tribe (e.g. barbarians) and in a game with all tribes. 
you may activate the europeans add on or a world addon as well to see and test

**Comment by frankystone** [comment]
Can you please merge master?

It's a bit difficult to understand what you mean.

For me loading a game made with master containing only Empire tribe shows in Tab 'Immovable' 4 entries for `Balsa Tree (old)`. The differences between those 4 entries are the preferred terrains. Same for 'Ironwood Tree (Old)', see image. Same happens with a new game and all 5 official tribes. 'Dead Tree' shows up in both, 'Immovables' and 'World Immovables' tabs.

Not sure if in a game the immovables are such importand to have so many help. E.g. i don't think that pebbles or the 'Dead Tree' are important for a game. At least immovables with `size = 0` are superfluous imho.

Apropos size:
The plot of grapevine seems to be wrong to me:

grapevine_medium_plot

Sorry for being critical... surely because i do not really understand the problem, or i never used the encyclopedia...

**Comment by hessenfarmer** [comment]
ok first I can confirm the wrong size of the grapevine. Will fix this in any case.     
Now I try to explain the reason for this attempt. Currently (in master) in the immovables tab all immovables are shown which can be used by the tribe (defined in their units.lua) . this includes e.g. all amazon trees with the helptexts that they are planted by amazons and can only be harvested. What is missing is the same information for all or at least all relevant world immovables. So i tried to implement this information to the encyclopedia. 
For the co-used immovables (amazon trees for all tribes, all flowering fields for frisians etc) this looks a bit weird to have some non tribe immovables (the co-used ones) in tribes tab and other (like normal trees and premature steps of the immovabled in the world section. 
so following options:
leave everything like it is closing the issue - result not all information abouit usable immovables available in encyclopedia (normal trees and rocks are missing)      
implement one of the solutions 1a to 1c  as described ion a previous post .

Another problem is that all amzon trees have 4 entries (for each former world summer, winter etc), where the difference is only terrain affinity to make them grow on any terrain

**Comment by frankystone** [comment]
Thanks for merging master :-)

Since i didn't used the in game encyclopedia yet i have to look at it and think about it in it's current state...

> leave everything like it is closing the issue - result not all information abouit usable immovables available in encyclopedia (normal trees and rocks are missing)

No, i think the issue is important, but looking at the current state of the encyclopdia in master i think there should be more general thoughts about the immovables in the encyclopedia... Just adding _all_ immovables isn't helpful for the player imho.

Maybe i am wrong... as often i do not really understand some things ;)

File: data/tribes/immovables/destroyed_building/init.lua [hunk 1]
```diff
@@ -7,7 +7,7 @@ descriptions:new_immovable_type {
    -- TRANSLATORS: This is an immovable name used in lists of immovables
    descname = pgettext("immovable", "Destroyed building"),
    icon = dirname .. "menu.png",
-   size = "big",
```
Comment (Noordfrees) [inline data/tribes/immovables/destroyed_building/init.lua]: :-1:
This change would have a big game-mechanical impact – you cannot build anything directly next to a Big immovable so currently you have to wait until the Destroyed Building turns to Ashes before you can build something new nearby

File: data/world/immovables/rocks/blackland/init.lua [hunk 2]
```diff
@@ -7,7 +7,7 @@ wl.Descriptions():new_immovable_type{
    descname = _ "Rocks 1",
    animation_directory = dirname,
    icon = dirname .. "menu1.png",
-   size = "big",
+   size = "small",
```
Comment (Noordfrees) [inline data/world/immovables/rocks/blackland/init.lua]: Same for all rocks

File: data/tribes/immovables/destroyed_building/init.lua (hunk 1 above)
Comment (hessenfarmer) [reply data/tribes/immovables/destroyed_building/init.lua]: actually the size of this immovable isn't properly taken into account. so I just adjusted the help to reflect real implementation. To verify build a big building in master, destroy it and watch the buildcaps appear of the formerly blocked 3 fields.

File: data/world/immovables/rocks/blackland/init.lua (hunk 2 above)
Comment (hessenfarmer) [reply data/world/immovables/rocks/blackland/init.lua]: rocks can be placed just necxt to each other so in fact they are not big immovables requireing big plots

File: data/tribes/immovables/destroyed_building/init.lua (hunk 1 above)
Comment (Noordfrees) [reply data/tribes/immovables/destroyed_building/init.lua]: The size does not have a game-logical connection to the number of fields occupied. The rule "4 nodes for Big, 1 node for every other size" applies only to buildings.

Oh, and workers (and other bobs) can move through immovables with size Small but not through immovables with size Medium or Big, so this too has an impact.

File: data/world/immovables/rocks/blackland/init.lua (hunk 2 above)
Comment (Noordfrees) [reply data/world/immovables/rocks/blackland/init.lua]: Calling the Size of an immovable a Plot Requirement in the help is actually misleading. All Immovables can always be placed on any node, regardless of their size and the node's buildcaps. Workers who create immovables may look specifically for nodes with a given minimum buildcaps though. 
The Size of immovables is taken into account for the computation of the buildcaps and movecaps of nearby fields, and whether the node is passable for bobs. 
For buildings only, the number of nodes they occupy depends on their Size, and buildings are never passable for bobs.

File: data/tribes/immovables/destroyed_building/init.lua (hunk 1 above)
Comment (hessenfarmer) [reply data/tribes/immovables/destroyed_building/init.lua]: actually i tested this on crater building a big building at ht esouthmost possibility west of the HQ.  After destroying it a big building plot became available north of it so I assumed the destroyed building in fact being a small building.  

maybe it should be a medium building, if workers really can move through small imovables. However I never saw any of them walk through a small building.

File: data/world/immovables/rocks/blackland/init.lua (hunk 2 above)
Comment (hessenfarmer) [reply data/world/immovables/rocks/blackland/init.lua]: ok so maybe rephrase this to size with adding the consequences textually like "can't be walked through by bobs"

File: data/tribes/immovables/destroyed_building/init.lua (hunk 1 above)
Comment (Noordfrees) [reply data/tribes/immovables/destroyed_building/init.lua]: This illustrates the difference between a Big immovable and a Small one (images from master):
grafik
grafik


> However I never saw any of them walk through a small building.

I did say below that "[…] buildings are never passable for bobs." :) 
Buildings are a very special case in many regards, and several of the rules that apply to all other immovables do not apply to them and vice versa.

Comment (hessenfarmer) [reply data/tribes/immovables/destroyed_building/init.lua]: ok got it. I just looked above the building and did not realize the change in the right. 
So I'll revert this

File: data/world/immovables/rocks/blackland/init.lua (hunk 2 above)
Comment (hessenfarmer) [reply data/world/immovables/rocks/blackland/init.lua]: implemented it, however I am unsure whether a shipconstruction siote is walkable or not.

File: data/tribes/scripting/help/immovable_help.lua [hunk 3]
```diff
@@ -55,8 +55,16 @@ function immovable_help_string(tribe, immovable_description)
       result = result .. h2(_"Requirements")
       if (buildcost ~= "") then
          result = result .. h3(_"Build cost:") .. buildcost
+         result = result .. plot_size_line(immovable_description.size)
+         result = result .. p(_"Workers and animals can't walk on fields with this immovable.")
```
Comment (Noordfrees) [inline data/tribes/scripting/help/immovable_help.lua]:
```suggestion
         result = result .. p(_"Workers and animals can’t walk on fields with this immovable.")
```

File: data/tribes/scripting/help/immovable_help.lua [hunk 4]
```diff
@@ -55,8 +55,16 @@ function immovable_help_string(tribe, immovable_description)
...
+      else
+         result = result .. plot_size_line(immovable_description.size, true)
+         if (immovable_description.size == "small") then
+            result = result .. p(_"Workers and animals can walk on fields with this immovable.")
+         else
+            result = result .. p(_"Workers and animals can't walk on fields with this immovable.")
```
Comment (Noordfrees) [inline data/tribes/scripting/help/immovable_help.lua]:
```suggestion
            result = result .. p(_"Workers and animals can’t walk on fields with this immovable.")
```

**Review by Noordfrees — COMMENTED** [review]
Shipconstructions are regular immovables with size Small so they are also passable. Also note that size "none" (or whatever it is called… the one that pebbles and dead trees have) is passable as well.

And perhaps clarify the hint a bit further by replacing "…can/can’t walk **on** fields with…" with "walk **across** fields" or "cross fields" or such – strictly speaking, they can *reach* such a node if it is the last step on their path (as stonemasons prove when walking to a Rock), they just can't walk *through* it (soldiers can walk right through a dense forest of trees but not through a barrier of rocks).

**Comment by hessenfarmer** [comment]
ok I reduced the list to immovables were at least any information is contained. 
I reduced the multiple entries in tribes immovables of amzaon trees, but this led to the terain affinity not being shown correctly.     
still todo:
create menu sized images as the currently used idle Pictures are too big for the help. 

@frankystone @Noordfrees would you consider the missing terrain affinity of the amazontrees multientries a bigger problem then having the multiple entries. 
Is the list of world immovables sufficient now.?

**Comment by frankystone** [comment]
Looks great :+1: 

Some things to consider:
1. I would suggest to rename the tool tipp for Immovables to `Tribe Immovables`. This makes it easier to understand the differences between 'Immovables' and 'World Immovables'.
2. Regarding the `Size` i am not sure if this information is helpful at all for immovables. E.g. `Rocks1` shows it needs a big Plot. One can think after 'harvesting' `Rocks1` he/she can place a big building here, which is not always true. For Rocks in general the information is missing that those can be 'harvested'. `Standing Stones` have also size big. In a game i think this information isn't useful: This stone is standing around and can't be removed. This information is useful in the editor, but in a game?
3. Some world immovables are missing: I think immovables of size  `none` can be omitted, but immovables like `Debris` or `Ruin` should be added because those can't be removed?
4. The trees in 'Immovables' are formatted different, e.g. `Becomes` → `Deadtree` shows an image. See image below. Because there are so many trees it may be good to have an own formatting-function for this type of immovable which can be used in `Immovables` and `World Immovables`?
5. The coloring of normal text in 'World Immovables' is different than in the other tabs. This should be consistent imho (this is also needed if text coloring get adjustable with a theme Add-On):
    encyclopedia_1

**Comment by hessenfarmer** [comment]
>     1. I would suggest to rename the tool tipp for Immovables to `Tribe Immovables`. This makes it easier to understand the differences between 'Immovables' and 'World Immovables'.

done.

>     2. Regarding the `Size` i am not sure if this information is helpful at all for immovables. E.g. `Rocks1` shows it needs a big Plot. One can think after 'harvesting' `Rocks1` he/she can place a big building here, which is not always true. For Rocks in general the information is missing that those can be 'harvested'. `Standing Stones` have also size big. In a game i think this information isn't useful: This stone is standing around and can't be removed. This information is useful in the editor, but in a game?

will have a look if we can just show Size: small without the word plot and without the buildhelp icon.ö so the info is reduced to walkability

>     3. Some world immovables are missing: I think immovables of size  `none` can be omitted, but immovables like `Debris` or `Ruin` should be added because those can't be removed?

in my build they are present, 
 
>     4. The trees in 'Immovables' are formatted different, e.g. `Becomes` → `Deadtree` shows an image. See image below. Because there are so many trees it may be good to have an own formatting-function for this type of immovable which can be used in `Immovables` and `World Immovables`?

this was due to missing images for the dead trees I corrected this
   
>     5. The coloring of normal text in 'World Immovables' is different than in the other tabs. This should be consistent imho (this is also needed if text coloring get adjustable with a theme Add-On):

This is only present if minimalistic Theme Addon is activated don't know why

**Comment by frankystone** [comment]
There are Richtext errors for:

 `Aspen` :
~~~~
WARNING: Error rendering richtext: [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/graphic/image_io.h:33] Image not found: world/immovables/trees/aspen/old/menu.png. Text is:
[00:10:39.998 real] WARNING: <rt><p><font face=sans size=18 color=fafafa bold=1>Aspen</font><br></p><vspace gap=6><p><font face=sans size=12 color=fafafa><img src=world/immovables/trees/aspen/old/../menu.png>...
~~~~
and `Oak`:
~~~~
WARNING: Error rendering richtext: [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/graphic/image_io.h:33] Image not found: world/immovables/trees/aspen/old/menu.png. Text is:
[00:09:56.051 real] WARNING: <rt><p><font face=sans size=18 color=fafafa bold=1>Oak</font><br></p><vspace gap=6><p><font face=sans size=12 color=fafafa><img src=world/immovables/trees/oak/old/../menu.png>....
~~~~

Regarding Debris/Ruin: Looks like world immovables are only shown if they are present on the map. That's fine, imho.

Starting a two players game Empire vs Atlanteans (Map: Together wer'e strong) the World Immovables shows Cassava Fields and Barley field. Starting a 8 Players game (Map: Dolomites) with all 5 tribes and some random tribes World Immovables doesn't show Cassava Fields and Barley fields.

Probably i have not enough knowledge of all the immovables to make good review.

Edit: This is Widelands Version 1.1~git25707[83826b0@3448_world_immovables] (Debug)

**Comment by hessenfarmer** [comment]
I will investigate about the cassava field issue. 
regarding the richtext it seems like there might be problems with the path being amended by adding ../before menu.png in your system. could you please check whether these errors asre present while using the editor tree help in master as well.     

About debris and ruins I tested this on crater and everything was present including all standing stones and artifacts. so something is not working as intended in your installtion of the branch. I will try to investigate however as this might be dependent on addons I would be interested:
1. in your exact addon Konfiguration
2. 2. if you could try to reproduce this with no addon activated as well

**Comment by frankystone** [comment]
This is Widelands Version 1.1~git25630[e9f047d@master] (Debug):

All trees are shown correct in the editors tree help, no richtext issues.

Currently i have only the minimalistic theme installed and enabled (i really like it ;) ).

This is Widelands Version 1.1~git25707[83826b0@3448_world_immovables] (Debug):

New game on map crater, using default players the World Immovable entries:
- contains one artifact
- no Debris
- no Ruins
- no standing stones
encyclopedia_2

Disabling the minimalistic theme does not change anything, the result is the same.

Edit: The richtext error applies only to this two trees, all other work fine.

**Comment by frankystone** [comment]
Strange is that selecting `Oak` the error reports also the not found image for Aspen...

> 
> ```
> WARNING: Error rendering richtext: [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/graphic/image_io.h:33] Image not found: world/immovables/trees/aspen/old/menu.png. Text is:
> [00:09:56.051 real] WARNING: <rt><p><font face=sans size=18 color=fafafa bold=1>Oak</font><br></p><vspace gap=6><p><font face=sans size=12 color=fafafa>....
> ```

**Comment by hessenfarmer** [comment]
ok I found the problem with oak and aspen.        

problem to solve is the difference in registered descriptions depending on whether an addon modyfying inits is activated or not.       
Background is that with such an addon all descriptions off all tribes and all world map objects get registered, while without only available world units and available tribes units are registered.      
@Noordfrees easiest way would to revert the selective loading and register all objects everytime. This would cope for consisten behaviour with and without addons

**Comment by Noordfrees** [comment]
> @Noordfrees easiest way would to revert the selective loading and register all objects everytime. This would cope for consisten behaviour with and without addons

The current behaviour is just a workaround for #4786, in the long run there will have to be dynamic loading with add-ons as well.

**Comment by hessenfarmer** [comment]
I found a way to remove other tribes fields from world immovables. so there should be only trees, rocks, ruins, debris and standing stones, if present on the map.
still to do:
fix size message (if possible)         Edit: done
test with scenarios                        Edit: done
find out why the text is shown differently in minimalistic theme        Edit: fixed

**Comment by hessenfarmer** [comment]
ok all issues fixed (I believe)

File: data/tribes/scripting/help/init.lua [hunk 5]
```diff
@@ -46,6 +57,23 @@ function immovable_entries(tribename)
...
+function world_immovable_entries(tribename)
+   local all_immovables = wl.Descriptions().immovable_descriptions
+   local tribes_immovables = {}
+   for i, tribe in ipairs(wl.Descriptions().tribes_descriptions) do
+      for i, t_immo in ipairs(tribe.immovables) do
+         tribes_immovables[t_immo.name] = true
+      end
+   end
+   for i = #all_immovables, 1, -1 do
+      if tribes_immovables[all_immovables[i].name] or all_immovables[i].size == "none" or (not all_immovables[i]:has_attribute("tree") and all_immovables[i].terrain_affinity) or all_immovables[i]:has_attribute("field") then
```
Comment (frankystone) [inline data/tribes/scripting/help/init.lua]: Can we have multiple line for this?

All other issues seems to be solved, except the coloring of text. Using Version git25712[1f4bcdf@3448_world_immovables] (Debug) and minimalistic theme the text colors for normal text in World Immovables is gray.

Comment (hessenfarmer) [reply data/tribes/scripting/help/init.lua]: done.     
the color worked and then I made a silly error. Now it should do

**Review by frankystone — APPROVED** [review]
Approved testing :-)

**Comment by hessenfarmer** [comment]
thanks for the review
